Implement `word-break: keep-all` (#9673)

flacerdk opened the pull request:
<!-- Please describe your changes on the following line: -->

Implement the `keep-all` value for the `word-break` property, as specified in [CSS](https://drafts.csswg.org/css-text-3/#word-break-property).

The relevant CSSWG tests (in `tests/wpt/css-tests/css-text-3_dev/html/word-break-keep-all-*.htm`) do not currently pass. As far as I can tell, this is because the tests use some JavaScript code that is not working properly. (But then, it seems that most tests in this directory are failing at the moment. I'm not sure what can be done here for now.)

---

<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #9673.

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/13414)

<!-- Reviewable:end -->

flacerdk committed:
Implement `word-break: keep_all`

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @jdm (or someone else) soon.

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/properties/longhand/inherited_text.mako.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style, layout, and gfx code, but no tests are modified. Please consider adding a test!

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 0c80d84 with merge 6675ea8...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/3172)

jdm commented:
@flacerdk Could you be more specific about the JS that's not working and how it is not doing so?

jdm commented:
http://build.servo.org/builders/linux-rel/builds/3348/steps/test_3/logs/stdio shows a lot of "FAIL [expected PASS]" results, but also a bunch of "PASS [expected FAIL]" for `/css-text-3_dev/html/css3-text-line-break-jazh-*` tests.

jdm commented:
Oh, I bet you're hitting #12939, since the tests are checking offsetWidth of a span :<

jdm commented:
r? @mbrubeck

flacerdk commented:
Indeed that seems to be the problem with the word-break tests.

I had overlooked the line-break ones; I'll take a look into that now.

flacerdk commented:
Oops, looks like I messed up with rebasing and accidentally closed this.

The line-break tests should pass now. There are some other tests that are not passing on my machine, but they also seem to be failing when I build HEAD, so hopefully it's just an issue with my local setup.

emilio commented:
@bors-servo: try

bors-servo commented:
:hourglass: Trying commit f16cff4 with merge 4c6cf8a...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3126), [arm64](http://build.servo.org/builders/arm64/builds/3141), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3209), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3386), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3195), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3186), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3321), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1090)

emilio commented:
I'm surprised no text expectations change, can you write a reftest that tests this if there isn't, please? Thanks! :)

jdm commented:
Note: docs on [writing tests](https://github.com/servo/servo/blob/master/tests/wpt/README.md#writing-new-tests) exist.

flacerdk commented:
I'll work on that; I suppose I can write versions of the currently existing tests that don't depend on #12939 being fixed.

jdm commented:
Certainly I don't think it's necessary to duplicate all of them, but a couple representative ones demonstrating the correct functioning of keep-all would be valuable.

mbrubeck commented:
Nice work! This got a bit more complicated than I initially thought it would be, but it looks like you figured it out.  Sorry I didn't catch up with this sooner to give some earlier feedback.

The approach here works, but I think it could be improved some. In particular, I think we should optimize for the default case (`word-break: normal`). This PR currently runs the line-breaking algorithm twice for the `normal` case: first during `break_and_shape` to find word boundaries, then again in `SoftWrapSliceIterator` to scan for breaking opportunities within each word. It would be nice to avoid this double scan.

You [noted correctly](https://github.com/servo/servo/issues/9673#issuecomment-249358884) that Servo currently expects each non-breakable slice to lie within a single GlyphRun. (This assumption is actually incorrect, but that's a somewhat separate issue; see #874.) If we want to keep the current behavior for `word-break: normal`, while also making sure that each word is a single GlyphRun for `word-break: keep-all`, then we need to change `TextRun::break_and_shape` to produce different GlyphRuns depending on the value of the `word-break` property. This could be done using a new flag in `gfx::font::ShapingFlags`. Then I think the existing `NaturalWordSliceIterator` could be used in both cases.

flacerdk committed:
Use word-break to decide how glyph runs should be created

flacerdk committed:
add tests

flacerdk commented:
I've made the suggested changes. Indeed it's much better to use the word-break info when creating the glyph runs.

I also added a few tests. Those are the same as the CSSWG ones, where I worked around the offsetWidth issue by using getBoundingClientRect() instead. I hope I added them in the right place! (Among those tests there's one for Thai which is currently failing; it also fails in every major browser except for Edge and IE.)

stshine commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 44c26e0 with merge 2355b90...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3419)

jdm commented:
That failure isn't related to these changes.

mbrubeck commented on components/layout/fragment.rs:
```diff
@@ -1655,6 +1655,15 @@ impl Fragment {
                     max_inline_size,
                     flags)
+            },
+            word_break::T::keep_all => {
```
It looks like this can now be combined with the `normal` case above.

flacerdk replied:
Yep, I just realized the same :) (is it OK to commit again?)

mbrubeck replied:
Oops, I didn't mean to post this comment. (I was waffling over whether it was better to keep these separate or not.)  But if you want to combine them, feel free.

mbrubeck commented:
@bors-servo r+

> I've made the suggested changes.

That was fast!  Thanks. :)

> Among those tests there's one for Thai which is currently failing; it also fails in every major browser except for Edge and IE.

For reference, Thai linebreaking is #13088.

bors-servo commented:
:pushpin: Commit 44c26e0 has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck 44c26e014aceec8ab45d419619b5f321779618f2 -->

mbrubeck commented:
@bors-servo r- delegate+

for minor cleanup (see commit comments above).  When this is ready to land, leave a comment starting with "@bors-servo r+".

bors-servo commented:
:v: @flacerdk can now approve this pull request

bors-servo commented:
:pushpin: Commit 44c26e0 has been approved by `mbrubeck".`

<!-- @bors-servo r=mbrubeck". 44c26e014aceec8ab45d419619b5f321779618f2 -->

bors-servo commented:
:v: @flacerdk can now approve this pull request

flacerdk committed:
combine normal and keep-all into a single strategy

flacerdk commented:
@bors-servo r=mbrubeck

bors-servo commented:
:pushpin: Commit fb66abb has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck fb66abba74df33d05eaf7a2a5e8ad8f2c88aff41 -->

bors-servo commented:
:hourglass: Testing commit fb66abb with merge e11d689...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3364)

mbrubeck commented:
@bors-servo retry

* #12654

bors-servo commented:
:hourglass: Testing commit fb66abb with merge 8f451cf...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3248)

KiChjang commented:
@bors-servo retry

flacerdk commented:
Did the buildbot die?

mbrubeck commented:
@bors-servo r+ retry

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

- There's another pull request that is currently being tested, blocking this pull request: #13323

bors-servo commented:
:pushpin: Commit fb66abb has been approved by `mbrubeck`

<!-- @bors-servo r=mbrubeck fb66abba74df33d05eaf7a2a5e8ad8f2c88aff41 -->

bors-servo commented:
:hourglass: Testing commit fb66abb with merge 2e65d8a...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/3484)

jdm commented:
@bors-servo: retry
* #13479

bors-servo commented:
:hourglass: Testing commit fb66abb with merge c944234...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/3223), [arm64](http://build.servo.org/builders/arm64/builds/3236), [linux-dev](http://build.servo.org/builders/linux-dev/builds/3308), [linux-rel](http://build.servo.org/builders/linux-rel/builds/3488), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/3300), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/3281), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/3425), [windows-dev](http://build.servo.org/builders/windows-dev/builds/1186)
